### PR TITLE
[dynamo] Skip dynamo when inside a functorch context

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118583
* #118616
* __->__ #118901

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by zou3519 — COMMENTED** [review]
This is going to merge conflict massively with https://github.com/pytorch/pytorch/pull/118407. Since that is adding more functorch support, while this PR is removing it, should we wait for that to land?

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -761,10 +762,24 @@ def clone_inputs(example_inputs):
...
+        raise SkipFrame(
+            f"torch.compile cannot be run in context: {functorch_subclass_name}"
+        ) from e
+
+
 @contextmanager
 def preserve_rng_state():
     with torch.utils._python_dispatch._disable_current_modes():
         rng_state = torch.clone(torch.random.get_rng_state())
+        skip_frame_if_in_functorch_mode(rng_state)
```
Comment (zou3519) [inline torch/_dynamo/utils.py:782]: Is there a better location to put this? (is this function always the first time tensor.data_ptr() gets called?)

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -761,10 +762,24 @@ def clone_inputs(example_inputs):
     return res
 
 
+def skip_frame_if_in_functorch_mode(val: torch.Tensor):
+    try:
+        val.data_ptr()  # will throw for functorch tensors
```
Comment (zou3519) [inline torch/_dynamo/utils.py:767]: functorch is not the only place where one can get a C++ tensor subclass without a data_ptr, but in general we should do what this PR does (graph break on Tensors that don't have data_ptr).

File: torch/_dynamo/utils.py (hunk 1 above)
Comment (jansel) [reply torch/_dynamo/utils.py:782]: This is a decorator around everything else, so it is the first call where issues happen.

File: torch/_dynamo/utils.py (hunk 2 above)
Comment (jansel) [reply torch/_dynamo/utils.py:767]: Yeah, they should all get graph breaks.

**Comment by jansel** [comment]
@zou3519 the rest of the stack is blocked on this PR, so I'd prefer not to wait.

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -761,10 +762,24 @@ def clone_inputs(example_inputs):
...
+    except RuntimeError as e:
+        from .exc import SkipFrame
+
+        # This will be GradTrackingTensor/BatchedTensor/etc
+        functorch_subclass_name = re.sub(r"\(.*", "", repr(val))
```
Comment (zou3519) [inline torch/_dynamo/utils.py:772]: @jansel, can you make this less functorch-centric? functorch is not the only place where one can get a Tensor with no data_ptr.

**Review by zou3519 — APPROVED** [review]

File: torch/_dynamo/utils.py (hunk 3 above)
Comment (jansel) [reply torch/_dynamo/utils.py:772]: This should work for all of them.  Though the functorch ones are the only ones that show up in tests.

**Pull request closed without merging**